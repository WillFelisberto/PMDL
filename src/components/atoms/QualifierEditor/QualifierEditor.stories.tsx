import type { Meta, StoryObj } from '@storybook/react';

import { QualifierEditor } from './QualifierEditor';

const meta: Meta<typeof QualifierEditor> = {
  title: 'Atoms/QualifierEditor',
  component: QualifierEditor,
  tags: ['autodocs'],
  args: {
    children: 'Example',
    variant: 'default'
  }
};

export default meta;

type Story = StoryObj<typeof QualifierEditor>;

export const Default: Story = {};
