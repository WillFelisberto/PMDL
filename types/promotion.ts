export type ConditionType =
  | 'comparator'
  | 'iterator'
  | 'group-iterator'
  | 'quantifier'
  | 'operator'
  | 'operation';

export type ConditionBase = {
  id: string;
  type: ConditionType;
  attributes?: Array<{ name: string; value: string }>;
};

export type ComparatorCondition = ConditionBase & {
  type: 'comparator';
  comparatorType:
    | 'equals'
    | 'not-equals'
    | 'less-than'
    | 'less-than-or-equals'
    | 'greater-than'
    | 'greater-than-or-equals'
    | 'contains'
    | 'starts-with'
    | 'ends-with'
    | 'isoneof'
    | 'isnotoneof'
    | 'includes'
    | 'includes-any'
    | 'includes-all';
  leftValue: string;
  rightValue: string;
  dataType: string;
};

export type ValueType = {
  type: 'value' | 'constant' | 'array' | 'null';
  dataType?: string;
  value?: string | number | boolean;
  values?: ValueType[];
};
export type IteratorCondition = ConditionBase & {
  type: 'iterator';
  iteratorType: 'next' | 'every' | 'up-to-and-including' | 'custom';
  collectionName: string;
  elementName: string;
  number?: number;
  sortBy?: string;
  sortOrder?: 'asc' | 'desc';
  condition?: PromotionCondition;
};

export type GroupIteratorCondition = ConditionBase & {
  type: 'group-iterator';
  iteratorType: 'every-group' | 'up-to-and-including-group' | 'custom';
  number?: number;
  conditions?: PromotionCondition[];
};

export type QuantifierCondition = ConditionBase & {
  type: 'quantifier';
  quantifierType: 'exactly' | 'at-least' | 'at-most' | 'all' | 'custom';
  number?: number;
  collectionName: string;
  elementName: string;
  condition?: PromotionCondition;
  customType?: string;
};

export type OperatorCondition = ConditionBase & {
  type: 'operator';
  operatorType: 'and' | 'or' | 'not' | 'custom';
  conditions: PromotionCondition[];
  customType?: string; // Para operadores personalizados
};

export type OperationCondition = ConditionBase & {
  type: 'operation';
  operationType: 'union' | 'anded-union' | 'custom';
  iterators: PromotionCondition[];
  customType?: string;
};
export type PromotionCondition =
  | ComparatorCondition
  | IteratorCondition
  | GroupIteratorCondition
  | QuantifierCondition
  | OperatorCondition
  | OperationCondition;

export type DiscountStructure = {
  calculatorType: string;
  discountType: 'percentOff' | 'amountOff' | 'fixedPrice';
  adjuster: number;
  target?: 'order' | 'shipping' | 'item';
};

export type Offer = {
  discountStructures: DiscountStructure[];
};

export type Promotion = {
  displayName: string;
  description: string;
  enabled: boolean;
  priority: number;
  qualifier: PromotionCondition | undefined;
  offer: Offer;
  conditions: PromotionCondition[];
  sites: Array<{ repositoryId: string }>;
};
