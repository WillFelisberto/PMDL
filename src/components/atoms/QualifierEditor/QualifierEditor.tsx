import { PromotionCondition, OperatorCondition, ComparatorCondition } from '@/types/promotion';

import { SortableCondition } from '@/components/atoms/SortableCondition';
import { SortableGroup } from '@/components/atoms/SortableGroup';

type QualifierEditorProps = {
  qualifier: PromotionCondition | undefined;
  onUpdate: (qualifier: PromotionCondition | undefined) => void;
};

export const QualifierEditor = ({ qualifier, onUpdate }: QualifierEditorProps) => {
  return (
    <div className="mb-6 p-4 border rounded-lg bg-gray-50">
      <h3 className="text-lg font-semibold text-gray-700">Qualificador</h3>

      {qualifier ? (
        <>
          {qualifier.type === 'operator' ? (
            <SortableGroup
              key={qualifier.id}
              group={qualifier}
              onUpdate={onUpdate}
              onDelete={() => onUpdate(undefined)}
            />
          ) : (
            <SortableCondition
              key={qualifier.id}
              condition={qualifier as ComparatorCondition}
              onUpdate={onUpdate}
              onDelete={() => onUpdate(undefined)}
            />
          )}
        </>
      ) : (
        <p className="text-gray-500 text-sm">Nenhuma condição adicionada.</p>
      )}

      {/* Botões para adicionar operadores (AND/OR) */}
      {!qualifier && (
        <div className="mt-4 flex gap-2">
          <button
            data-testid="add-and-group"
            onClick={() => {
              const newQualifierGroup: OperatorCondition = {
                id: crypto.randomUUID(),
                type: 'operator',
                operatorType: 'and',
                conditions: []
              };
              onUpdate(newQualifierGroup);
            }}
            className="bg-blue-500 text-white px-4 py-2 rounded-md hover:bg-blue-600 transition-colors"
          >
            + Adicionar Grupo (AND)
          </button>

          <button
            data-testid="add-or-group"
            onClick={() => {
              const newQualifierGroup: OperatorCondition = {
                id: crypto.randomUUID(),
                type: 'operator',
                operatorType: 'or',
                conditions: []
              };
              onUpdate(newQualifierGroup);
            }}
            className="bg-green-500 text-white px-4 py-2 rounded-md hover:bg-green-600 transition-colors"
          >
            + Adicionar Grupo (OR)
          </button>

          <button
            onClick={() => {
              const newQualifier: ComparatorCondition = {
                id: crypto.randomUUID(),
                type: 'comparator',
                comparatorType: 'equals',
                leftValue: '',
                rightValue: '',
                dataType: 'java.lang.String'
              };
              onUpdate(newQualifier);
            }}
            data-testid="add-simple-condition"
          >
            + Adicionar Condição Simples
          </button>
        </div>
      )}
    </div>
  );
};
