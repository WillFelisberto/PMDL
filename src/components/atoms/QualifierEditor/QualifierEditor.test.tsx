import { QualifierEditor } from './QualifierEditor';

import { render, screen, fireEvent } from '@/tests/test-utils';
import { OperatorCondition, ComparatorCondition } from '@/types/promotion';

// Mock para crypto.randomUUID
beforeAll(() => {
  Object.defineProperty(global, 'crypto', {
    value: {
      randomUUID: () => 'mock-uuid-1234'
    }
  });
});

describe('QualifierEditor', () => {
  const mockOnUpdate = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('deve exibir botões de adição quando não houver qualificador', () => {
    const { container } = render(<QualifierEditor qualifier={undefined} onUpdate={mockOnUpdate} />);

    expect(screen.getByTestId('add-and-group')).toBeInTheDocument();
    expect(screen.getByTestId('add-or-group')).toBeInTheDocument();
    expect(screen.getByTestId('add-simple-condition')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('deve adicionar um grupo AND ao clicar no botão', () => {
    render(<QualifierEditor qualifier={undefined} onUpdate={mockOnUpdate} />);

    fireEvent.click(screen.getByTestId('add-and-group'));

    expect(mockOnUpdate).toHaveBeenCalledWith(
      expect.objectContaining({
        type: 'operator',
        operatorType: 'and'
      })
    );
  });

  it('deve exibir e remover uma condição simples', async () => {
    const condition: ComparatorCondition = {
      id: '1',
      type: 'comparator',
      comparatorType: 'equals',
      leftValue: 'order.total',
      rightValue: '100',
      dataType: 'java.lang.Double'
    };

    render(<QualifierEditor qualifier={condition} onUpdate={mockOnUpdate} />);

    // Hover para mostrar o botão de delete
    const conditionElement = screen.getByTestId('sortable-condition');
    fireEvent.mouseEnter(conditionElement);

    const deleteButton = await screen.findByTestId('delete-button');
    fireEvent.click(deleteButton);

    expect(mockOnUpdate).toHaveBeenCalledWith(undefined);
  });

  it('deve exibir e interagir com um grupo AND', async () => {
    const group: OperatorCondition = {
      id: '2',
      type: 'operator',
      operatorType: 'and',
      conditions: []
    };

    render(<QualifierEditor qualifier={group} onUpdate={mockOnUpdate} />);

    // Verificar elementos do grupo
    expect(screen.getByTestId('sortable-group')).toBeInTheDocument();
    expect(screen.getByTestId('add-condition-button')).toBeInTheDocument();
    expect(screen.getByTestId('add-group-button')).toBeInTheDocument();

    // Testar hover para mostrar botão de delete
    const groupElement = screen.getByTestId('sortable-group');
    fireEvent.mouseEnter(groupElement);

    const deleteButton = await screen.findByTestId('delete-button');
    fireEvent.click(deleteButton);
    expect(mockOnUpdate).toHaveBeenCalledWith(undefined);
  });

  it('deve adicionar uma nova condição dentro do grupo', async () => {
    const group: OperatorCondition = {
      id: '3',
      type: 'operator',
      operatorType: 'and',
      conditions: []
    };

    render(<QualifierEditor qualifier={group} onUpdate={mockOnUpdate} />);

    fireEvent.click(screen.getByTestId('add-condition-button'));

    expect(mockOnUpdate).toHaveBeenCalledWith(
      expect.objectContaining({
        conditions: [
          expect.objectContaining({
            type: 'comparator'
          })
        ]
      })
    );
  });

  it('deve atualizar valores de uma condição', async () => {
    let currentCondition: ComparatorCondition = {
      id: '4',
      type: 'comparator',
      comparatorType: 'equals',
      leftValue: '',
      rightValue: '',
      dataType: 'java.lang.String'
    };

    const { rerender } = render(
      <QualifierEditor
        qualifier={currentCondition}
        onUpdate={(updated) => {
          currentCondition = updated as ComparatorCondition;
          mockOnUpdate(updated);
        }}
      />
    );

    // Atualizar e verificar leftValue
    fireEvent.change(screen.getByTestId('left-value-input'), {
      target: { value: 'order.paymentMethod' }
    });
    expect(mockOnUpdate).toHaveBeenLastCalledWith(
      expect.objectContaining({
        leftValue: 'order.paymentMethod',
        rightValue: ''
      })
    );

    // Atualizar e verificar rightValue
    rerender(<QualifierEditor qualifier={currentCondition} onUpdate={mockOnUpdate} />);
    fireEvent.change(screen.getByTestId('right-value-input'), { target: { value: 'pix' } });

    expect(mockOnUpdate).toHaveBeenLastCalledWith(
      expect.objectContaining({
        leftValue: 'order.paymentMethod',
        rightValue: 'pix'
      })
    );
  });
});
