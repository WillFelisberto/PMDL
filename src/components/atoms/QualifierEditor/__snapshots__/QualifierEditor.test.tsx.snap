// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QualifierEditor deve exibir botões de adição quando não houver qualificador 1`] = `
<div>
  <div
    class="mb-6 p-4 border rounded-lg bg-gray-50"
  >
    <h3
      class="text-lg font-semibold text-gray-700"
    >
      Qualificador
    </h3>
    <p
      class="text-gray-500 text-sm"
    >
      Nenhuma condição adicionada.
    </p>
    <div
      class="mt-4 flex gap-2"
    >
      <button
        class="bg-blue-500 text-white px-4 py-2 rounded-md hover:bg-blue-600 transition-colors"
        data-testid="add-and-group"
      >
        + Adicionar Grupo (AND)
      </button>
      <button
        class="bg-green-500 text-white px-4 py-2 rounded-md hover:bg-green-600 transition-colors"
        data-testid="add-or-group"
      >
        + Adicionar Grupo (OR)
      </button>
      <button
        data-testid="add-simple-condition"
      >
        + Adicionar Condição Simples
      </button>
    </div>
  </div>
</div>
`;
